File: backend/src/main/java/com/kratosgado/chatapp/utils/ApiException.java
package com.kratosgado.chatapp.utils;

import org.springframework.http.HttpStatus;
import lombok.Getter;

@Getter
public class ApiException extends RuntimeException {
  private final HttpStatus status;

  public ApiException(HttpStatus status, String message) {
    super(message);
    this.status = status;
  }

  public static ApiException badRequest(String message) {
    throw new ApiException(HttpStatus.BAD_REQUEST, message);
  }

  public static ApiException unauthorized(String message) {
    throw new ApiException(HttpStatus.UNAUTHORIZED, message);
  }

  public static ApiException forbidden(String message) {
    throw new ApiException(HttpStatus.FORBIDDEN, message);
  }

  public static ApiException notFound(String message) {
    throw new ApiException(HttpStatus.NOT_FOUND, message);
  }

  public static ApiException conflict(String message) {
    throw new ApiException(HttpStatus.CONFLICT, message);
  }

  public static ApiException unprocessableEntity(String message) {
    throw new ApiException(HttpStatus.UNPROCESSABLE_ENTITY, message);
  }

  public static ApiException internalServerError(String message) {
    throw new ApiException(HttpStatus.INTERNAL_SERVER_ERROR, message);
  }

  public static ApiException serviceUnavailable(String message) {
    throw new ApiException(HttpStatus.SERVICE_UNAVAILABLE, message);
  }

}

File: backend/src/main/kotlin/com/kratosgado/chatapp/utils/ApiException.kt
package com.kratosgado.chatapp.utils

import org.springframework.http.HttpStatus

class ApiException(
    val status: HttpStatus,
    message: String,
) : RuntimeException(message) {
    companion object {
        fun badRequest(message: String) = ApiException(HttpStatus.BAD_REQUEST, message)

        fun unauthorized(message: String) = ApiException(HttpStatus.UNAUTHORIZED, message)

        fun forbidden(message: String) = ApiException(HttpStatus.FORBIDDEN, message)

        fun notFound(message: String) = ApiException(HttpStatus.NOT_FOUND, message)

        fun conflict(message: String) = ApiException(HttpStatus.CONFLICT, message)

        fun unprocessableEntity(message: String) = ApiException(HttpStatus.UNPROCESSABLE_ENTITY, message)

        fun internalServerError(message: String) = ApiException(HttpStatus.INTERNAL_SERVER_ERROR, message)

        fun serviceUnavailable(message: String) = ApiException(HttpStatus.SERVICE_UNAVAILABLE, message)
    }
}
